src/pybind/mgr/dashboard/controllers/auth.py

```diff
@@ -186,7 +224,39 @@ def check(self, token):
                     'sso': mgr.SSO_DB.protocol == 'saml2',
                     'pwdUpdateRequired': user.pwd_update_required
                 }
-        return {
-            'login_url': self._get_login_url(),
-            'cluster_status': ClusterModel.from_db().dict()['status']
-        }
+        elif hasattr(cherrypy.request, 'cookie') and '_oauth2_proxy' in cherrypy.request.cookie and \
+          cherrypy.request.cookie['_oauth2_proxy'].value:
+            orch = OrchClient()
+            if not orch.services.list_daemons(daemon_type='mgmt-gateway'):
+                logger.info('mgmt-gateway is not configured, can not access Dashboard with oauth2 SSO')
+                return default_return()
+            access_token = cherrypy.request.headers.get('X-Access-Token')
+            if not access_token:
+                return default_return()
+            JwtManager.oauth2_token = True
+            token_payload = json.loads(base64.urlsafe_b64decode(access_token.split(".")[1] + "===="))
+            if not 'sub' in token_payload:
+                return default_return()
+            try:
+                user = AuthManager.get_user(token_payload['sub'])
+            except UserDoesNotExist:
+                user = mgr.ACCESS_CTRL_DB.create_user(token_payload['sub'], None, token_payload['name'], token_payload['email'])
+                if 'resource_access' in token_payload:
+                    # Find the first value where the key is not 'account'
+                    user_roles = next(value for key, value in token_payload['resource_access'].items() if key != "account")
+                    if user_roles:
+                        user.set_roles([Role.map_to_system_roles(user_roles['roles'][0])])
+                else: #if no valid role is provided set readonly as default
+                    user.set_roles([READ_ONLY_ROLE])
+                user.last_update=token_payload['iat'] #set user last update to token time issued
+                mgr.ACCESS_CTRL_DB.save()
+            if user:
+                url_prefix = 'https' if mgr.get_localized_module_option('ssl') else 'http'
+                self._set_token_cookie(url_prefix, access_token)
+                return {
+                    'username': user.username,
+                    'permissions': user.permissions_dict(),
+                    'sso': mgr.SSO_DB.protocol == 'saml2',
+                    'pwdUpdateRequired': user.pwd_update_required
+                }
```

Comment:
this function clearly handle the case of `oauth2-proxy`, I'd recommend having something like (please, read the code and modify it carefully .. I din't test it, just a re-factorization suggestion):

```
    def parse_user_roles(token_payload):
        # first we cehck for Client Roles section
        if 'resource_access' in token_payload:
            # Find the first value where the key is not 'account'
            user_roles = next(value for key, value in token_payload['resource_access'].items() if key != "account")
            return [Role.map_to_system_roles(user_roles['roles'][0])] if ('roles' in user_roles and user_roles['roles']) else []
        # if it doesn't exist we cehck for Realm Roles section
         elif 'resource_access':
             return ..

     def get_user(token_payload):
        try:
            user = AuthManager.get_user(token_payload['sub'])
        except UserDoesNotExist:
            user = mgr.ACCESS_CTRL_DB.create_user(token_payload['sub'], None, token_payload['name'], token_payload['email'])
            user_roles = parse_user_roles(token_payload)
            if user_roles:
                user.set_roles(user_roles)
            else:
                user.set_roles([READ_ONLY_ROLE])
            user.last_update=token_payload['iat'] #set user last update to token time issued
            mgr.ACCESS_CTRL_DB.save()
        return user

    @RESTController.Collection('POST', query_params=['token'])
    @EndpointDoc("Check token Authentication",
                 parameters={'token': (str, 'Authentication Token')},
                 responses={201: AUTH_CHECK_SCHEMA})
    def check(self, token):
        JwtManager.oauth2_token = False

        def default_return():
           return {
                    'login_url': self._get_login_url(),
                    'cluster_status': ClusterModel.from_db().dict()['status']
                }

        if token:
            user = JwtManager.get_user(token)
            if user:
                return {
                    'username': user.username,
                    'permissions': user.permissions_dict(),
                    'sso': mgr.SSO_DB.protocol == 'saml2',
                    'pwdUpdateRequired': user.pwd_update_required
                }
        elif is_oidc_request(cherrypy.request):

            orch = OrchClient()
            if not orch.services.list_daemons(daemon_type='mgmt-gateway'):
                logger.info('mgmt-gateway is not configured, can not access Dashboard with oauth2 SSO')
                return default_return()

            access_token = cherrypy.request.headers.get('X-Access-Token')
            if not access_token:
                return default_return()
            JwtManager.oauth2_token = True
            token_payload = json.loads(base64.urlsafe_b64decode(access_token.split(".")[1] + "===="))
            if not 'sub' in token_payload:
                return default_return()

            user = get_user()
            url_prefix = 'https' if mgr.get_localized_module_option('ssl') else 'http'
            self._set_token_cookie(url_prefix, access_token)
            return {
                'username': user.username,
                'permissions': user.permissions_dict(),
                'sso': mgr.SSO_DB.protocol == 'saml2',
                'pwdUpdateRequired': user.pwd_update_required
            }

        return default_return()

```
I would also look into adding the part: 

```
        if token:
            user = JwtManager.get_user(token)
            if user:
                return {
                    'username': user.username,
                    'permissions': user.permissions_dict(),
                    'sso': mgr.SSO_DB.protocol == 'saml2',
                    'pwdUpdateRequired': user.pwd_update_required
                }
```
into the get user ... as it has some which being repeated later .. but that's more risky in as modification because it changes the behaviour and the lines:

``` 
           url_prefix = 'https' if mgr.get_localized_module_option('ssl') else 'http'
            self._set_token_cookie(url_prefix, access_token)
``` 

are only called for oauth2-proxy case...

Reply:
Thanks Redo, guided by your suggestions I refactored the code